fdbserver/RestoreApplier.actor.cpp

```diff
@@ -152,313 +150,280 @@ ACTOR static Future<Void> handleSendMutationVectorRequest(RestoreSendVersionedMu
 	}
+	return Void();
+}
 
-	// Has all mutations been committed?
-	bool isDone() { return curItInCurTxn == batchData->kvOps.end(); }
-
-	// Set cursor for next mutation
-	void nextMutation() {
-		curIndexInCurTxn++;
-		while (curItInCurTxn != batchData->kvOps.end() && curIndexInCurTxn >= curItInCurTxn->second.size()) {
-			curIndexInCurTxn = 0;
-			curItInCurTxn++;
-			startNextVersion = true;
+// Get keys in imcompleteStagingKeys and precompute the stagingKey which is stored in batchData->stagingKeys
+ACTOR static Future<Void> getAndComputeStagingKeys(
+    std::map<Key, std::map<Key, StagingKey>::iterator> imcompleteStagingKeys, Database cx, UID applierID) {
+	state Reference<ReadYourWritesTransaction> tr(new ReadYourWritesTransaction(cx));
+	state std::vector<Future<Optional<Value>>> fValues;
+	state std::vector<Optional<Value>> values;
+	state int i = 0;
+	TraceEvent("FastRestoreApplierGetAndComputeStagingKeysStart", applierID)
+	    .detail("GetKeys", imcompleteStagingKeys.size());
+	loop {
+		try {
+			tr->reset();
+			tr->setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
+			tr->setOption(FDBTransactionOptions::LOCK_AWARE);
+			for (auto& key : imcompleteStagingKeys) {
+				fValues.push_back(tr->get(key.first));
+			}
+			for (i = 0; i < fValues.size(); i++) {
+				Optional<Value> val = wait(fValues[i]);
+				values.push_back(val);
+			}
+			break;
+		} catch (Error& e) {
+			TraceEvent(SevError, "FastRestoreApplierGetAndComputeStagingKeysUnhandledError")
+			    .detail("GetKeys", imcompleteStagingKeys.size())
+			    .detail("Error", e.what())
+			    .detail("ErrorCode", e.code());
+			wait(tr->onError(e));
+			fValues.clear();
 		}
 	}
 
-	// Setup for the next transaction; This should be done after nextMutation()
-	void nextTxn() {
-		txnBytes = 0;
-		txnMutations = 0;
-		numAtomicOps = 0;
-		lastTxnHasError = false;
-		startNextVersion = false;
+	ASSERT(values.size() == imcompleteStagingKeys.size());
+	int i = 0;
+	for (auto& key : imcompleteStagingKeys) {
+		if (!values[i].present()) {
+			TraceEvent(SevWarnAlways, "FastRestoreApplierGetAndComputeStagingKeysUnhandledError")
+			    .detail("Key", key.first)
+			    .detail("Reason", "Not found in DB")
+			    .detail("PendingMutations", key.second->second.pendingMutations.size())
+			    .detail("StagingKeyType", (int)key.second->second.type);
+			for (auto& vm : key.second->second.pendingMutations) {
+				for (auto& m : vm.second) {
+					TraceEvent(SevWarnAlways, "FastRestoreApplierGetAndComputeStagingKeysUnhandledError")
+					    .detail("PendingMutationVersion", vm.first)
+					    .detail("PendingMutation", m.toString());
+				}
+			}
+			key.second->second.precomputeResult();
+			i++;
+			continue;
+		} else {
+			// The key's version ideally should be the most recently committed version.
+			// But as long as it is > 1 and less than the start version of the version batch, it is the same result.
+			MutationRef m(MutationRef::SetValue, key.first, fValues[i].get().get());
```

Comment:
You can uses `values[i].get()` here. I guess `values` can be removed, and you can just use `fValues[i].get()` anywhere `values[i]` is used.